### PR TITLE
JAMES-2186 GetMessages delegation

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature [hunk 1]
```diff
@@ -17,6 +17,24 @@ Feature: GetMessages method
...
+    Given "alice@domain.tld" shares its mailbox "shared" with "bob@domain.tld"
+    And the user has a message "m1" in "shared" mailbox with subject "my test subject", content "testmail"
+    Given "bob@domain.tld" is connected
+    When the user ask for messages "m1"
+    Then no error is returned
+    And the list should contain 1 message
+    And the id of the message is "m1"
+
+  Scenario: Retrieving a message in a mailbox not delegated to me
+    Given "alice@domain.tld" has a mailbox "shared"
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: what about naming it notshared?
what about adding a test where the mailbox is shared, but with another user (let's name here Carol)

Comment (rouazana) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: PS: if one of these tests is failing as stated in this commit, instead put it in ignore (that's a tag)

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java [hunk 2]
```diff
@@ -90,9 +95,8 @@ public StoreMessageIdManager(MailboxSessionMapperFactory mailboxSessionMapperFac
     @Override
     public void setFlags(Flags newState, MessageManager.FlagsUpdateMode replace, MessageId messageId, List<MailboxId> mailboxIds, MailboxSession mailboxSession) throws MailboxException {
         MessageIdMapper messageIdMapper = mailboxSessionMapperFactory.getMessageIdMapper(mailboxSession);
-        MailboxMapper mailboxMapper = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession);
 
-        allowOnMailboxSession(mailboxIds, mailboxSession, mailboxMapper);
+        assertRightsOnMailboxes(mailboxIds, mailboxSession, MailboxACL.Right.Write);
```
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java]: maybe you can import MailboxACL.Right

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java [hunk 3]
```diff
@@ -103,10 +107,9 @@ public void setFlags(Flags newState, MessageManager.FlagsUpdateMode replace, Mes
     @Override
     public Set<MessageId> accessibleMessages(Collection<MessageId> messageIds, MailboxSession mailboxSession) throws MailboxException {
         MessageIdMapper messageIdMapper = mailboxSessionMapperFactory.getMessageIdMapper(mailboxSession);
-        MailboxMapper mailboxMapper = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession);
         List<MailboxMessage> messageList = messageIdMapper.find(messageIds, MessageMapper.FetchType.Metadata);
         return messageList.stream()
-            .filter(message -> mailboxBelongsToUser(mailboxSession, mailboxMapper).test(message.getMailboxId()))
+            .filter(hasRightsOn(mailboxSession, MailboxACL.Right.Read))
```
Comment (rouazana) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java]: no unit test on these changes?

**Comment by mbaechler** [comment]
test this please

**Comment by mbaechler** [comment]
test this please

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature (hunk 1 above)
Comment (chibenwa) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/GetMessages.feature]: About the order: It's github ordering: the branch introduce integretion tests at last (but they were 1st thing developed)

**Comment by strokyl** [comment]
This PR now incorporate commit from this other PR: https://github.com/linagora/james-project/pull/1039

File: mailbox/api/src/test/java/org/apache/james/mailbox/model/Rfc4314RightsTest.java [hunk 4]
```diff
@@ -212,4 +212,19 @@ public void unionShouldReturnOriginWhenAppliedWithEmpty() throws UnsupportedRigh
...
+
+    @Test
+    public void containsShouldReturnFalseWhenAtLeastOneRightNotPresent() throws UnsupportedRightException {
+        assertThat(lprs.contains(Lookup, Write)).isFalse();
+    }
+
+    @Test
+    public void containsShouldReturnTrueWhenAllRightsPresent() throws UnsupportedRightException {
+        assertThat(lprs.contains(Lookup, Post)).isTrue();
+    }
```
Comment (aduprat) [inline mailbox/api/src/test/java/org/apache/james/mailbox/model/Rfc4314RightsTest.java]: Can you add a test when calling _contains_ without parameter (empty varargs).

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/InMemoryMessageIdManager.java [hunk 5]
```diff
@@ -140,18 +140,12 @@ public void setInMailboxes(MessageId messageId, List<MailboxId> mailboxIds, Mail
     }
 
     private List<MailboxId> getUsersMailboxIds(MailboxSession mailboxSession) throws MailboxException {
-        return mailboxManager.search(userMailboxes(mailboxSession), mailboxSession)
+        return mailboxManager.search(MailboxQuery.builder().matchesAllMailboxNames().build(), mailboxSession)
```
Comment (aduprat) [inline mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/InMemoryMessageIdManager.java]: What about a _MailboxQuery.ALL_MAILBOXES_QUERY_ constant?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java [hunk 6]
```diff
@@ -34,12 +34,14 @@
 import javax.mail.Flags;
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageIdManager.java]: Detailed test of this changeset will be provided after rebasing https://github.com/linagora/james-project/pull/1042

File: server/protocols/jmap/src/test/java/org/apache/james/jmap/DefaultMailboxesProvisioningFilterThreadTest.java [hunk 7]
```diff
@@ -199,6 +199,11 @@ public boolean hasRight(MailboxPath mailboxPath, MailboxACL.Right right, Mailbox
         }
 
         @Override
+        public MailboxACL.Rfc4314Rights myRights(MailboxId mailboxId, MailboxSession session) throws MailboxException {
+            return null;
```
Comment (chibenwa) [inline server/protocols/jmap/src/test/java/org/apache/james/jmap/DefaultMailboxesProvisioningFilterThreadTest.java]: Throw NotImplemented(blah)

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/InMemoryMessageIdManager.java (hunk 5 above)
Comment (quynhn) [reply mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/InMemoryMessageIdManager.java]: I don't think it make sense, it's clear with builder IMO

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadGet.feature [hunk 8]
```diff
@@ -61,4 +62,22 @@ Feature: Download GET
     And a connected user "username1@domain.tld"
     And "username1@domain.tld" has a mailbox "INBOX"
     When "username1@domain.tld" downloads "1"
-    Then the user should receive a not found response
\ No newline at end of file
+    Then the user should receive a not found response
+
+  Scenario: User cannot download attachment of another user when shared mailbox
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadGet.feature]: cannot? it seems he can

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadGet.feature [hunk 9]
```diff
@@ -61,4 +62,22 @@ Feature: Download GET
...
+    Given "username@domain.tld" mailbox "sharedMailbox" contains a message "1" with an attachment "2"
+    And "username@domain.tld" shares its mailbox "sharedMailbox" with "username1@domain.tld"
+    And a connected user "username1@domain.tld"
+    And "username1@domain.tld" has a mailbox "sharedMailbox"
+    When "username1@domain.tld" downloads "2"
+    Then the user should receive that blob
+    And the blob size is 3071
+
+  Scenario: User cannot download message blob of another user when shared mailbox
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadGet.feature]: idem

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadGet.feature [hunk 10]
```diff
@@ -61,4 +62,22 @@ Feature: Download GET
...
+    Given "username@domain.tld" mailbox "sharedMailbox" contains a message "1" with an attachment "2"
+    And "username@domain.tld" shares its mailbox "sharedMailbox" with "username1@domain.tld"
+    And a connected user "username1@domain.tld"
+    And "username1@domain.tld" has a mailbox "sharedMailbox"
+    When "username1@domain.tld" downloads "1"
+    Then the user should receive that blob
+    And the blob size is 4963
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/DownloadGet.feature]: can you add 2 tests where alice share with bob, and carol cannot access this share, one for attachment and one for message blob?
